src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,393 @@
+    */
+  private def createDataIterator(inputRows: Iterator[Row], numPassesInSpark: Int) = {
+    if (numPassesInSpark == 1)
+    // avoid caching inputRows
+      () => inputRows
+    else {
+      // cache all row references to be able to quickly get the iterator again
+      val inputRowsArr = inputRows.toArray
+      () => inputRowsArr.iterator
+    }
+  }
+
+  private def buildCommandLineArguments(vwArgs: String, contextArgs: () => String = () => "") = {
+    val args = new StringBuilder
+    args.append(vwArgs)
+      .append(" ").append(contextArgs())
+
+    // have to pass to get multi-pass to work
+    if (args.indexOf("--no_stdin") == -1)
+      args.append(" --no_stdin")
+
+    // need to keep reference around to prevent GC and subsequent file delete
+    if (getEnableCacheFile) {
+      val cacheFile = java.io.File.createTempFile("vowpalwabbit", ".cache")
+      cacheFile.deleteOnExit()
+
+      args.append(s" -k --cache_file=${cacheFile.getAbsolutePath} --passes ${getNumPasses}")
+    }
+
+    log.info(s"VowpalWabbit args: ${args}")
+
+    args.result
+  }
+
+  /**
+    * Internal training loop.
+    * @param df the input data frame.
+    * @param vwArgs vw command line arguments.
+    * @param contextArgs This lambda returns command line arguments that are executed in the final execution context.
+    *                    It is used to get the partition id.
+    */
+  private def trainInternal(df: DataFrame, vwArgs: String, contextArgs: () => String = () => "") = {
+    val applyLabel = createLabelSetter(df)
+
+    val featureColIndices = generateNamespaceInfos(df.schema)
+
+    // only perform the inner-loop if we cache the inputRows
+    val numPassesInSpark = if (getEnableCacheFile) 1 else if(getCacheRows) getNumPasses else 1
+
+    def trainIteration(inputRows: Iterator[Row],
+                       localInitialModel: Option[Array[Byte]],
+                       pass: Int): Iterator[Option[Array[Byte]]] = {
+
+      // potential cache data
+      val iteratorGenerator = createDataIterator(inputRows, numPassesInSpark)
+
+      // construct command line arguments
+      val args = buildCommandLineArguments(vwArgs, contextArgs)
+
+      StreamUtilities.using(if (localInitialModel.isEmpty) new VowpalWabbitNative(args)
+        else new VowpalWabbitNative(args, localInitialModel.get)) { vw =>
+        StreamUtilities.using(vw.createExample()) { ex =>
+
+            // loop in here to avoid
+            // - passing model back and forth
+            // - re-establishing network connection
+            for (_ <- 0 until numPassesInSpark) {
+              // pass data to VW native part
+              for (row <- iteratorGenerator()) {
+                // transfer label
+                applyLabel(row, ex)
+
+                // transfer features
+                for (ns <- featureColIndices)
+                  row.get(ns.colIdx) match {
+                    case dense: DenseVector => ex.addToNamespaceDense(ns.featureGroup,
```

Comment:
can VW handle mixed sparse and dense vectors?  spark has that often for different rows.

Reply:
I had to convert to one type in lightgbm

Reply:
VW internally treats everything sparse. Mapping from dense to sparse is straight forward in this layer.